### PR TITLE
add redux store, finished login screen,layout,routing

### PR DESCRIPTION
**Pull request by NazarChopko**

### CODE REVIEW

File: src/components/DashboardHandlers.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import Button from "@mui/material/Button";
+import AddIcon from "@mui/icons-material/Add";
+
+const DashboardHandlers = () => {
+  const dispatch = useCustomDispatch();
+  const navigate = useNavigate();
+
+  const addNewVacation = () => {
+    dispatch(handleBackButton(true));
+    dispatch(editTitle("Add vacation"));
```
Comment (Avinamor) [inline src/components/DashboardHandlers.tsx]: you don't need redux for component title or back button

File: src/components/DatePicker.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import { AdapterDayjs } from "@mui/x-date-pickers/AdapterDayjs";
+import { LocalizationProvider } from "@mui/x-date-pickers/LocalizationProvider";
+import { DatePicker } from "@mui/x-date-pickers/DatePicker";
+
+interface IVacationDatePickerProps {
+  type: string;
+  startDate?: Dayjs | null;
+  endDate?: Dayjs | null;
+  setStartDate?: any;
+  setEndDate?: any;
```
Comment (Avinamor) [inline src/components/DatePicker.tsx]: please don't use `any`

File: src/components/DatePicker.tsx [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+const VacationDatePicker: FC<IVacationDatePickerProps> = ({
+  type,
+  startDate,
+  endDate,
+  setEndDate,
+  setStartDate,
+}) => {
+  const handleChangeDate = (date: any) => {
```
Comment (Avinamor) [inline src/components/DatePicker.tsx]: please don't use `any`

File: src/components/Layout.tsx [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+import Typography from "@mui/material/Typography";
+import Menu from "@mui/material/Menu";
+import Avatar from "@mui/material/Avatar";
+import Tooltip from "@mui/material/Tooltip";
+import MenuItem from "@mui/material/MenuItem";
+import ListItemIcon from "@mui/material/ListItemIcon";
+import Logout from "@mui/icons-material/Logout";
+import ArrowBackIcon from "@mui/icons-material/ArrowBack";
+
+function ResponsiveAppBar() {
```
Comment (Avinamor) [inline src/components/Layout.tsx]: const ResponsiveAppBar =

File: src/components/Layout.tsx [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+  const [openUserSettings, setOpenUserSettings] = useState(null);
+  const { user, logout, loading } = useAuth();
+  const { isBackButton, title } = useTypedSelector(
+    (state) => state.controlSlice
+  );
+  const dispatch = useCustomDispatch();
+  const navigate = useNavigate();
+
+  const handleOpenUserMenu = (event: any) => {
```
Comment (Avinamor) [inline src/components/Layout.tsx]: don't use `any`

File: src/components/Layout.tsx [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    setOpenUserSettings(event.currentTarget);
+  };
+
+  const handleCloseUserMenu = () => {
```
Comment (Avinamor) [inline src/components/Layout.tsx]: please add a return type for functions also. here it will be `void`

File: src/components/Layout.tsx [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+              textAlign: "center",
+            }}
+          >
+            {title}
+          </Typography>
+          <Box>
+            <Tooltip title="Open settings">
+              <IconButton onClick={handleOpenUserMenu} sx={{ p: 0 }}>
+                <Avatar
+                  alt={!loading && user ? user.email[0].toUpperCase() : "-"}
```
Comment (Avinamor) [inline src/components/Layout.tsx]: it can a method

File: src/pages/AddVacationPage.tsx [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+import VacationSelector from "../components/VacationSelector";
+
+// interface MyValues {
+//   type: string;
+//   start: string;
+//   end: boolean | any;
+// }
+
+const AddVacation = () => {
+  const [vacationType, setVacationType] = useState<string | null | undefined>(
```
Comment (Avinamor) [inline src/pages/AddVacationPage.tsx]: too much types

File: src/pages/Form.tsx [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+import Box from "@mui/material/Box";
+import Button from "@mui/material/Button";
+import TextField from "@mui/material/TextField";
+import Checkbox from "@mui/material/Checkbox";
+import FormControlLabel from "@mui/material/FormControlLabel";
+import Typography from "@mui/material/Typography";
+import { useAuth } from "../hooks/useAuth";
+import { useNavigate } from "react-router-dom";
+
+interface MyValues {
```
Comment (Avinamor) [inline src/pages/Form.tsx]: it's a bad naming for an interface

File: src/pages/Form.tsx [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+  email: string;
+  password: string;
+  remember: boolean | any;
```
Comment (Avinamor) [inline src/pages/Form.tsx]: why `any` is here?

File: src/redux/slices/controlSlice.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
+import { createSlice, PayloadAction } from "@reduxjs/toolkit";
+import { IControl, DashboardTitle } from "./types";
+
+const initialState: IControl = {
+  isBackButton: false,
+  title: "Dashboard",
```
Comment (Avinamor) [inline src/redux/slices/controlSlice.ts]: you don't need redux for hading those values

**Comment by Avinamor** [comment]
1. Each component needs to place in a folder. This folder name has to be the same as the component. Also, it's a good practice to add an index.ts file with a default export of the component.
2. Styles are better place in a separate file in the same folder as the component.
3. DatePicker might be a good example of a component for the "components" folder. Components specific to a section UI are better be allocated in a subfolder of this component.
4. there are some other points, but let's start with those ones.
Screenshot 2022-12-07 at 11 57 09

File: src/router/AppRouter.tsx [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+import { AddVacation } from "../pages/AddVacation";
+
+const AppRouter = () => {
+  const { user, loading } = useAuth();
+  const [isBackButton, setIsBackButton] = useState<boolean>(false);
+  const [title, setTitle] = useState<string>("Dashboard");
+
+  if (loading) return <div>loading</div>;
+
+  return user ? (
```
Comment (Avinamor) [inline src/router/AppRouter.tsx]: The router shouldn't depend on the user

File: src/router/AppRouter.tsx [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+import { Layout } from "../shared/Layout";
+import { useAuth } from "../hooks/useAuth";
+import { AddVacation } from "../pages/AddVacation";
+
+const AppRouter = () => {
+  const { user, loading } = useAuth();
+  const [isBackButton, setIsBackButton] = useState<boolean>(false);
+  const [title, setTitle] = useState<string>("Dashboard");
+
+  if (loading) return <div>loading</div>;
```
Comment (Avinamor) [inline src/router/AppRouter.tsx]: Loading here is not needed. It's handled on the components level.
Why do you think it should be here?

File: src/hooks/useAuth.tsx [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+  remember: boolean;
+}
+
+export const useAuth = () => {
+  const [user, setUser] = useState<User>({} as User);
+  const [loading, setLoading] = useState<boolean>(true);
+  const navigate = useNavigate();
+
+  useEffect(() => {
+    const isUser: User | null = JSON.parse(
```
Comment (Avinamor) [inline src/hooks/useAuth.tsx]: wrong naming. var name starting from `is` means it's a boolean, but you have User | null. Rename please

File: src/hooks/useAuth.tsx [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+import { useNavigate } from "react-router-dom";
+
+export interface User {
+  email: string;
+  password: string;
+  remember: boolean;
+}
+
+export const useAuth = () => {
+  const [user, setUser] = useState<User>({} as User);
```
Comment (Avinamor) [inline src/hooks/useAuth.tsx]: replace {} with null

File: src/hooks/useAuth.tsx [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+  const [loading, setLoading] = useState<boolean>(true);
```
Comment (Avinamor) [inline src/hooks/useAuth.tsx]: it should be false by default

File: src/pages/AddVacation/AddVacation.tsx [hunk 17]
```diff
@@ -0,0 +1,62 @@
+import React, { useState } from "react";
+import { Dayjs } from "dayjs";
+import Box from "@mui/material/Box";
+import TextField from "@mui/material/TextField";
+import { VacationDatePicker } from "../../shared/DatePicker";
+import { VacationSelector } from "../../shared/VacationSelector";
+import { styledAddVacation } from "./style";
+
+const AddVacation = () => {
+  const [vacationType, setVacationType] = useState<string | null>("");
```
Comment (Avinamor) [inline src/pages/AddVacation/AddVacation.tsx]: just string

File: src/pages/AddVacation/AddVacation.tsx [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+  const [startDate, setStartDate] = useState<Dayjs | null>(null);
+  const [endDate, setEndDate] = useState<Dayjs | null>(null);
+  const [notes, setNotes] = useState<string | null>();
```
Comment (Avinamor) [inline src/pages/AddVacation/AddVacation.tsx]: just string and default value ""

File: src/pages/AddVacation/AddVacation.tsx [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+  const [error, setError] = useState<string>("");
+
+  const isFildesEmpty = startDate && endDate && vacationType;
```
Comment (Avinamor) [inline src/pages/AddVacation/AddVacation.tsx]: typo in isFildesEmpty

File: src/pages/AddVacation/AddVacation.tsx [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+    if (!isFildesEmpty) return console.log("Error");
+    console.log({ vacationType, endDate, startDate });
+  };
+
+  return (
+    <>
+      <Box sx={styledAddVacation.vacationWrapper}>
+        <Box>
+          <Box sx={styledAddVacation.datePickBlock}>
+            <VacationSelector
```
Comment (Avinamor) [inline src/pages/AddVacation/AddVacation.tsx:31]: Put the VacationSelector component as a subfolder in the AddVacation folder. This component is specific only for AddVacation.

File: src/pages/Form/Form.tsx [hunk 21]
```diff
@@ -0,0 +1,96 @@
+import { useFormik, FormikProps } from "formik";
```
Comment (Avinamor) [inline src/pages/Form/Form.tsx]: rename folder, file according to the function name

File: src/router/AppRouter.tsx [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+import { UserForm } from "../pages/Form";
+import { Dashboard } from "../pages/Dashboard";
+import { Layout } from "../shared/Layout";
+import { useAuth } from "../hooks/useAuth";
+import { AddVacation } from "../pages/AddVacation";
+
+const AppRouter = () => {
+  const { user, loading } = useAuth();
+  const [isBackButton, setIsBackButton] = useState<boolean>(false);
+  const [title, setTitle] = useState<string>("Dashboard");
```
Comment (Avinamor) [inline src/router/AppRouter.tsx]: I didn't get the trick with isBackButton and title. It definitely shouldn't be here, in the Router.

It looks like Layout.tsx is just a header component?

So the idea is next: let's say you have a Dashboard component. It consists of a Header, filters/controls and table/calendar parts. The header is also used in other components. The only difference is in the title text and back-btn. 
So you need to create a stateless component Header.tsx that takes 2 props: title and backBtnAction. 

Now you can use Header in Dashboard and pass title="dashboard" without backBtnAction (which means it won't be displayed).
And in AddVacation.tsx you can use Header also. Just by passing title="Add Vacation"(or Edit, AddVacation will handle that) and backBtnAction={navigation.back()}

File: src/shared/DashboardHandlers/DashboardHandlers.tsx [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+import { useNavigate } from "react-router-dom";
+
+import Box from "@mui/material/Box";
+import ButtonGroup from "@mui/material/ButtonGroup";
+import Button from "@mui/material/Button";
+import AddIcon from "@mui/icons-material/Add";
+import { DashboardProps } from "../../pages/Dashboard/Dashboard";
+import { styledDashboard } from "./style";
+
+const DashboardHandlers: FC<DashboardProps> = ({
```
Comment (Avinamor) [inline src/shared/DashboardHandlers/DashboardHandlers.tsx]: I'd suggest renaming it something like DashBoardControls.

File: src/shared/DashboardHandlers/DashboardHandlers.tsx [hunk 24]
```diff
@@ -0,0 +1,49 @@
...
+  setTitle,
+  setIsBackButton,
```
Comment (Avinamor) [inline src/shared/DashboardHandlers/DashboardHandlers.tsx]: After the refactoring of AppRouter those props you won't need any more.

File: src/shared/Layout/Layout.tsx [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+import { styledLayout } from "./style";
+
+export interface ILayoutProps {
+  title: string;
+  isBackButton: boolean;
+  setTitle: Dispatch<SetStateAction<string>>;
+  setIsBackButton: Dispatch<SetStateAction<boolean>>;
+}
+
+const Layout: FC<ILayoutProps> = ({
```
Comment (Avinamor) [inline src/shared/Layout/Layout.tsx]: it's a part of refactoring that I wrote above

File: src/shared/Table/Table.tsx [hunk 26]
```diff
@@ -0,0 +1,209 @@
...
+  count: number;
+  page: number;
+  rowsPerPage: number;
+  onPageChange: (
+    event: React.MouseEvent<HTMLButtonElement>,
+    newPage: number
+  ) => void;
+}
+
+function TablePaginationActions(props: TablePaginationActionsProps) {
```
Comment (Avinamor) [inline src/shared/Table/Table.tsx:29]: folder and file name has to be the same as the function name

File: src/shared/VacationSelector/VacationSelector.tsx [hunk 27]
```diff
@@ -0,0 +1,41 @@
+import React, { FC, SetStateAction, Dispatch } from "react";
+import Box from "@mui/material/Box";
+import InputLabel from "@mui/material/InputLabel";
+import MenuItem from "@mui/material/MenuItem";
+import FormControl from "@mui/material/FormControl";
+import Select, { SelectChangeEvent } from "@mui/material/Select";
+
+interface IVacationSelectorProps {
+  vacationType: any;
```
Comment (Avinamor) [inline src/shared/VacationSelector/VacationSelector.tsx]: don't use `any`

**Review by Avinamor — CHANGES_REQUESTED** [review]

File: src/pages/UserForm/UserForm.tsx [hunk 28]
```diff
@@ -0,0 +1,96 @@
...
+import Box from "@mui/material/Box";
+import Button from "@mui/material/Button";
+import TextField from "@mui/material/TextField";
+import Checkbox from "@mui/material/Checkbox";
+import FormControlLabel from "@mui/material/FormControlLabel";
+import Typography from "@mui/material/Typography";
+import { useAuth } from "../../hooks/useAuth";
+import { styledForm } from "./style";
+
+interface IFormData {
```
Comment (Avinamor) [inline src/pages/UserForm/UserForm.tsx]: IUserFormData

File: src/router/AppRouter.tsx [hunk 29]
```diff
@@ -0,0 +1,17 @@
+import { Routes, Route, Navigate } from "react-router-dom";
+import { UserForm } from "../pages/UserForm";
+import { Dashboard } from "../pages/Dashboard";
+import { AddVacation } from "../pages/AddVacation";
+
+const AppRouter = () => {
+  return (
+    <Routes>
+      <Route path="/user" element={<Dashboard />} />
```
Comment (Avinamor) [inline src/router/AppRouter.tsx]: I guess this route should be "/"

File: src/router/AppRouter.tsx [hunk 30]
```diff
@@ -0,0 +1,17 @@
...
+      <Route path="/user/vacation" element={<AddVacation />} />
+      <Route path="/" element={<UserForm />} />
```
Comment (Avinamor) [inline src/router/AppRouter.tsx]: and this one - "/login"

File: src/shared/DashBoardControls/DashBoardControls.tsx [hunk 31]
```diff
@@ -0,0 +1,43 @@
...
+import ButtonGroup from "@mui/material/ButtonGroup";
+import Button from "@mui/material/Button";
+import AddIcon from "@mui/icons-material/Add";
+import { styledDashboard } from "./style";
+
+const DashBoardControls: FC = () => {
+  const navigate = useNavigate();
+
+  const addNewVacation = (): void => {
+    navigate("/user/vacation", { replace: true });
```
Comment (Avinamor) [inline src/shared/DashBoardControls/DashBoardControls.tsx]: Why do you use this prop for navigation?
{ replace: true }

Comment (NazarChopko) [reply src/shared/DashBoardControls/DashBoardControls.tsx]: I was wrong about how it prop works. My bad)

**Review by Avinamor — APPROVED** [review]

**Pull request merged**